File: 백준/Gold/16234. 인구 이동/인구 이동.py
import sys
sys.setrecursionlimit(10 ** 6)
#sys.stdin = open('input.txt', 'r')
input = sys.stdin.readline

n, l, r = map(int, input().split())
countries = [list(map(int, input().split())) for _ in range(n)]

# 동남서북 (시계 방향)
dx = [1, 0, -1, 0]
dy = [0, -1, 0, 1]


def dfs(y, x, visited, union):
    for d in range(4):
        ny, nx = y + dy[d], x + dx[d]
        if 0 <= ny < n and 0 <= nx < n and not visited[ny][nx]:
            difference = abs(countries[ny][nx] - countries[y][x])  # 차이
            if l <= difference <= r:  # 범위 이내에 있으면
                visited[ny][nx] = True  # 방문처리
                union.append((ny, nx))  # 연합에 저장
                dfs(ny, nx, visited, union)
    return


days = 0
while True:
    is_moved = False
    visited = [[False] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                union = [(i, j)] # 연합 초기화
                visited[i][j] = True # 현재 위치 방문처리
                dfs(i, j, visited, union) # dfs 시작
                if len(union) > 1:  # 연합이 하나 이상
                    avg = sum(countries[y][x] for y, x in union) // len(union)  # 평균 계산
                    for y, x in union:
                        countries[y][x] = avg  # 인구 이동 완료
                    is_moved = True # 인구이동 했음
    if not is_moved:  # 움직이지 않으면
        break

    days += 1 # 인구 이동이 일어난 경우에만 날짜 증가
print(days)

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.util.*;
import java.io.*;

public class Main {
    static int n,l,r;
    static int[][] map;
    static boolean[][] visited;
    static int[] dx = {0,0,-1,1};
    static int[] dy = {-1,1,0,0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());

        map = new int[n][n];
        for (int i = 0; i < n; i++){
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j ++){
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int days = 0;
        while (true){
            visited = new boolean[n][n];
            boolean moved = false;

            for (int i = 0; i < n; i ++){
                for (int j = 0; j < n; j ++){
                    if(!visited[i][j]){
                        if (bfs(i,j)) moved = true;

                    }
                }
            }

            if (!moved) break;
            days++;
        }
        System.out.println(days);
    }

    static boolean bfs(int x, int y){
        Queue<int[]> q = new LinkedList<>();
        List<int[]> union = new ArrayList<>();
        q.add(new int[]{x,y});
        union.add(new int[]{x,y});
        visited[x][y] = true;
        int sum = map[x][y];

        while (!q.isEmpty()){
            int[] curr = q.poll();
            int cx = curr[0], cy = curr[1];

            for (int d = 0; d < 4; d ++) {
                int nx = cx + dx[d];
                int ny = cy + dy[d];

                if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[nx][ny]){
                    int diff = Math.abs(map[cx][cy] - map[nx][ny]);
                    if(diff >= l && diff <= r){
                        visited[nx][ny] = true;
                        q.add(new int[]{nx,ny});
                        union.add(new int[]{nx,ny});
                        sum += map[nx][ny];
                    }
                }
            }
        }

        if (union.size() == 1) return false; // 연합이 없음
        int newPopulation = sum / union.size(); // 연합의 인구수 / 연합을 이루고 있는 칸의 갯수
        // 인구수 조정
        for (int[] cell: union){
            map[cell[0]][cell[1]] = newPopulation;
        }
        return true;
    }

}